### PR TITLE
Optionally install Ansible Filters for Sprig

### PR DESCRIPTION
**Pull request by ryandgoulding**
Sprig Functions are Go template function abstractions used to
manipulate date in Go templating.  They are extremely useful for tasks
such as data format manipulation and entity creation (i.e., passwords
or certificates).  A fuller explanation of Sprig can be found here:

http://masterminds.github.io/sprig/

This change abstracts an internal package called "filters".  This
package is responsible for placing some stub Ansible Filters into the
generated Ansible Playbook skeleton in order to aid in transition from
Helm (Go Templating) to Ansible Playbook (Jinja2).  Notably, there are
two filters installed when "generateFilters" flag is true:
1) sprig_ansible_filters.py:  A complete stubbing of Ansible Playbook
Filters set up to mimic Sprig Function method signatures.  Right now,
every filter is instructed to raise NotImplementedError.
2) invoke_go_tf:  This is a more advanced set of filters aimed to demo
how one could wrap Go Template functions in order to provide a partial
replacement.  These filters demonstrate invoking the Go runtime from an
Ansible Filter context, and utilizing stdout to reference the rendered
template.  Right now, this is fairly limited as data inputs are mapped
from strings based on template function method signature (best effort)
and output is limited to strings.

This is not meant to be an all encompassing change, but rather just a
bit of tooling to help users in the conversion process.

Signed-off-by: Ryan Goulding <rgouldin@redhat.com>

### CODE REVIEW

File: internal/filters/main.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+const funcArgsCLIArgsFirstIndex = 2
+const funcNameCLIArgsIndex = 1
+const funcReturnOutputIndex = 0
+
+// Wrapper to call a Golang function through reflection.
+func CallFunc(funcMap map[string]interface{}, name string, arguments ... interface{}) (result []reflect.Value,
+	err error) {
+
+	function := reflect.ValueOf(funcMap[name])
+	if len(arguments) != function.Type().NumIn() {
```
Comment (aneeshkp) [inline internal/filters/main.go:36]: NumIn panics if  the type's Kind is not Func. You should test .Type().NumIn()   and try to recover from panic and return gracefully and use golang recover() to catch the panic. May be you do this in the main function where you bind the function.
func CallFunc()....{
defer func() {
        if e := recover(); e != nil {
            err = errors.New(name + " is not a callable fucntion.")
        }
    }()
   function := reflect.ValueOf(funcMap[name])
   function.Type().NumIn()

File: internal/filters/main.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+func main() {
+	args := os.Args
+	// This would indicate a function with no funcName argument.
+	checkForFuncName(&args)
+
+	funcName := args[funcNameCLIArgsIndex]
+	funcArgs := args[funcArgsCLIArgsFirstIndex:]
+
+	// Loads only the text/template and Sprig Function Map.
+	funcMap := sprig.TxtFuncMap()
```
Comment (aneeshkp) [inline internal/filters/main.go:84]: I would put this in its own file and not have logic embeded in main.
Define a type for funcs like `type funcs map[string]reflect.Value` and then add methods to it.But if you are planning to throw this away and just for demo purpose, I am okay with that.

File: internal/filters/main.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+	// Used to convert CLI args into functional arguments.
+	var iFuncArgs []interface{}
+	f := reflect.ValueOf(funcMap[funcName])
+	var argType = reflect.String
+	for i, b := range funcArgs {
+		// Attempt to derive the argument type from the method signature, and cast/convert the input appropriately.
+		if i < f.Type().NumIn() {
```
Comment (aneeshkp) [inline internal/filters/main.go]: I tested it , it panics here, so add that recover logic that I mentioned above,  here . Once its already checked here you don;t have to do in CallFunc

File: internal/filters/main.go (hunk 2 above)
Comment (ryandgoulding) [reply internal/filters/main.go:84]: I agree with you in principal, but I would like to take the route that this is "throw away demo code".  If our stakeholders like what we show them, then we can consider replacing this with a more robust solution, such as https://github.com/tsg/gotpl.  This was very much a first pass, and we can discuss further investment with David, Andrew and others.

File: internal/filters/main.go (hunk 1 above)
Comment (ryandgoulding) [reply internal/filters/main.go:36]: I have an updated patch which does Kind checking to ensure that the **Value** is a **Func**.

File: internal/filters/main.go (hunk 3 above)
Comment (ryandgoulding) [reply internal/filters/main.go]: I resolved the panic by checking that the type is in fact a Function in a subsequent patch, per my comment above.  Additionally, I added error checking for failed casting.

This conditional is actually still needed;  this logic is attempting to cast the input arguments from **string** to match whatever is specified by the method signature.  However, some methods take variable length arguments, so there is certainly a point at which the CLI string argument index will exceed whatever is provided by **f.Type().In(i)**.  In this case, we know that whatever the argument type was uncovered during the last valid call to **f.Type.In(i)** will be the argument type for the rest of the var args.

**Review by aneeshkp — APPROVED** [review]
look good for now. Need to change if you make this solution  perm

**Pull request merged**